### ISSUE TITLE
release v5

### ISSUE THREAD

**Issue by Tieske** [homieiot/convention#299]
How about it? Imho the spec is ready.

**Comment by helix55** [comment]
That's a remarkable job. Is there a library which provides support for Homie v5?

**Comment by jacoscaz** [comment]
I'm working on a control platform for Homie V5 but it's going to take a few more months.

**Comment by Tieske** [comment]
@helix55 I'm working on a Lua lib (doesn't have child-devices yet). And I think @schaze has a JS lib, and possibly a Rust one?

**Comment by Tieske** [comment]
@Thalhammer what do we need to get the spec released? any ideas?

Also: I acquired the `homieiot.org` domain, so maybe we can use that as the entrypoint as well.

**Comment by schaze** [comment]
> @helix55 I'm working on a Lua lib (doesn't have child-devices yet). And I think @schaze has a JS lib, and possibly a Rust one?

Yes I am working on the rust library for v5 (mostly completed)- but not yet released or or available in github yet. It is a bit low level but due to this it can be used basically everywhere. I have implementations already publishing a device on an ESP32 SoC as well as a zigbee bridge controller and a TUI homie5 client which can browse, search and control v5 devices.

The NodeJS lib is still in beta. I did not update it for a quite some time. It is the same library as the v4 one, just with version 5 tags: https://github.com/schaze/node-homie/tree/5.0.0-beta5 - the beta5 is also released on npmjs.com already. But it is still missing some of latest updates done by @Tieske.

**Comment by helix55** [comment]
> > @helix55 I'm working on a Lua lib (doesn't have child-devices yet). And I think @schaze has a JS lib, and possibly a Rust one?
> 
> Yes I am working on the rust library for v5 (mostly completed)- but not yet released or or available in github yet. It is a bit low level but due to this it can be used basically everywhere. I have implementations already publishing a device on an ESP32 SoC as well as a zigbee bridge controller and a TUI homie5 client which can browse, search and control v5 devices.
> 
> The NodeJS lib is still in beta. I did not update it for a quite some time. It is the same library as the v4 one, just with version 5 tags: https://github.com/schaze/node-homie/tree/5.0.0-beta5 - the beta5 is also released on npmjs.com already. But it is still missing some of latest updates done by @Tieske.

Wow! This Rust library is greatly anticipated. The low-level implementation feature is highly appealing and precisely matches our usage scenarios. Perhaps we can add a Python wrapper to your Rust implementation if needed, which would enable compatibility with multiple language scenarios and enhance ease of use.

**Comment by schaze** [comment]
@helix55: Just a note - I have now released v0.1.0 of the homie5 rust library on crates.io: https://crates.io/crates/homie5
Feel free to check it out - but keep in mind this is early beta and a lot of things might still change. Documentation is still lacking but I put in 2 examples (device and controller) that more or less touch on everything.

**Comment by jacoscaz** [comment]
I do not know what sort of timeframe you were looking at for finalising / releasing V5. However, should there be interest in the concept, I think V5 could do with at least some foundational work on device profiles. Actual profiles can be added at a later point within minor / patch versions of the spec but the general concept of a device profile and what it means for a device to declare implementation for a specific profile are major features, worthy of inclusion in a major revision such as V5 (or V6 if V5 needs to come out very soon).

EDIT: forgot to link to #319 which I've opened to gauge interest in the idea.

**Comment by dcj** [comment]
As I have [just written elsewhere](https://github.com/homieiot/convention/issues/320#issuecomment-2764630688), I have adopted Homie 5 for a proposed integration framework, and I want to "note for the record" that due to Homie 5's current status as not "complete/finished/finalized" **and even worse (for me)** is referred to [here](https://homieiot.github.io/specification/)  as `develop` with no indication at that level that `develop` is `v5 in development` this has caused confusion, concern, and questions with potential business (integration) partners.

My constructive minimal request is to change `develop` to `v5-develop` or something similar, to reduce the level of confusion when I tell someone to go use Homie 5 and they say "that doesn't exist".  In other words, I am only asking that the `develop` label be agumented to indicate it is intended to become `v5` (someday).  

It would be super helpful to me if the Homie team were to provide some notion of the goals for v5, maybe a list of "stuff that needs to be done to finalize v5", or some guesstimated timeline for when that might happen, but of course you are under no obligation to do so.  If there are "things that need to done to finalize v5", some of those tasks are things that the team might welcome other volunteers to contribute, then perhaps articulating that might prompt others (maybe even me) to pitch in.  Some teams do not welcome or accept outside contributions, and that is absolutely their perogative.

Some of my other work involves the Clojure ecosystem/community, and I aspire to live-by/up-to the "[Clojure Etiquette](https://clojure.org/community/etiquette)" a key part of which is "don't be an entitled-acting user, if you want something done, do it yourself, don't ask/expect others to spend their time to do stuff for you", so I am (hopefully) not doing that here.

**Comment by schaze** [comment]
@Tieske I think there is nothing holding us back to release v5 into "production". What do you think?

**Comment by Tieske** [comment]
The one thing would be the stricter definition of the `type` attribute. Because making things stricter is a breaking change.

**Comment by Tieske** [comment]
> The one thing would be the stricter definition of the `type` attribute. Because making things stricter is a breaking change.

nvm, see [this comment](https://github.com/homieiot/convention/issues/320#issuecomment-2790520147) so I think we're good to go.

**Comment by jacoscaz** [comment]
If we're good to go, and I do think we are, how do we "push the button"? Can I help with anything? 

Also, it might be a good idea to do some spring cleaning of the open issues here on GitHub to better convey that the project is indeed active. A few examples:

- https://github.com/homieiot/convention/issues/287 could be addressed with a non-normative (or normative?) note in the convention documenting the approach suggested by @Thalhammer in https://github.com/homieiot/convention/issues/287#issuecomment-2190770750 . This could be done before V5, happy to craft a PR for it.
- https://github.com/homieiot/convention/issues/209 could be closed as the discussion appears to have come to its natural end.
- https://github.com/homieiot/convention/issues/247 appears to have been addressed by https://github.com/homieiot/convention/pull/253 and thus could also be closed.

**Comment by Tieske** [comment]
closed 2, the streams thing is an addition that can be added later. I'd prefer to get v5 shipped as it is now.

Just don't know how to properly do that

**Comment by jacoscaz** [comment]
>  I'd prefer to get v5 shipped as it is now.

Sounds good to me, minor changes can be addressed in minor revisions (if needed), anyway.

> Just don't know how to properly do that

Would it be worth setting up a meeting with perhaps @davidgraeff for some knowledge transfer on the best way to release V5 and other maintenance issues?

Insofar as I understand, based on https://github.com/homieiot/homieiot.github.io/blob/042d87d9a762b5ceb3b546a99f93a1947c355593/grabrepos.py#L4-L7 it should be enough to create a new `v5.0.0` git tag, perhaps mapped to an equivalent `v5.0.0` commit for clarity. This should trigger a new build of the https://homieiot.github.io website. During the build, the https://github.com/homieiot/homieiot.github.io/blob/master/grabrepos.py script will automatically grab all tags and render the appropriate version-specific pages in https://homieiot.github.io/specification/ . 

One change I would make is to display the the most recently released version of the spec when one clicks on the `specification` link rather than the version taken from the `develop` version branch. That's what most people are looking for, particularly given that a new version of the spec might take years to be finalised. I _think_ this is where this change should be made, though I haven't actually tested it and only had a cursory glance at the script: https://github.com/homieiot/homieiot.github.io/blob/042d87d9a762b5ceb3b546a99f93a1947c355593/grabrepos.py#L118

EDIT: we should create a `v5.0.0` tag, a `v5.0.0` commit and a `v5.0.0` GitHub release as per https://github.com/homieiot/convention/releases

**Comment by davidgraeff** [comment]
Yes, creating a tag is enough to release a new version. For future development, it might make sense to create -rc tags to indicate in-development revisions as requested by @dcj

**Comment by jacoscaz** [comment]
Trying to facilitate releasing V5, the pending tasks that I see are:

- [x] merge https://github.com/homieiot/convention/pull/329
- [ ] merge https://github.com/homieiot/homieiot.github.io/pull/34
- [ ] create new `v5.0.0` commit and tag 

Thinking of active members of the homeiot organization, merging should mainly be up to @schaze , I believe, as @Tieske has authored and/or reviewed both already. The commit and tag can be created by either.

Does anyone see anything else that would be strictly required for v5?

Then, tasks to be carried out immediately post-release:

- [ ] The introduction and device validation stuff on the site is all based on v4. So that would need an update as well (from @tieske's comment https://github.com/homieiot/convention/issues/299#issuecomment-2869965108 )

As per my previous comment, _after releasing v5_ I think a bit of spring cleaning of open issues could also help. Some of it might result in minor edits to be considered for a minor release (v5.1 or so). I'd be happy to help in this regard, opening PRs and/or pointing out things that can be closed.

**Comment by dcj** [comment]
After v5 release, I hope/plan to do a PR that enables URL anchor links on the section headers within the specification.  I want/need to be able to have URLs that resolve to the sections detailing specific items, e.g. "description", "property", "units", etc.

**Comment by jacoscaz** [comment]
@dcj if I understand correctly what you're after, that should already work.

Examples from current version:

- https://homieiot.github.io/specification/#devices
- https://homieiot.github.io/specification/#property-attributes

Examples from develop version:

- https://homieiot.github.io/specification/spec-core-develop/#units
- https://homieiot.github.io/specification/spec-core-develop/#device-attributes

**Comment by Tieske** [comment]
The introduction and device validation stuff on the site is all based on v4. So that would need an update as well

**Comment by jacoscaz** [comment]
@Tieske I've added that to the list in https://github.com/homieiot/convention/issues/299#issuecomment-2869945899 . Would you be happy with that list going up into a new tracking-style issue that would supersede this one? Just so that we could have the constantly-updated list as the first comment.

Separately, I think what you suggest can be split as follows to facilitate moving ahead, otherwise we risk entering a catch-22:

1. temporarily add a note to the website, pointing out that examples are outdates
2. release v5
3. update examples and website

**Comment by Tieske** [comment]
Homie 5 is tagged and released. Closing this, follow ups in #330
